provers/sp1/build.rs

```diff
@@ -117,30 +150,87 @@ fn build_program_with_dependencies(
     built_programs.insert(program.to_string());
 }
 
-// fn is_build_enabled() -> bool {
-//     env::var("SKIP_GUEST_BUILD").is_err() && env::var("CARGO_CFG_CLIPPY").is_err()
-// }
-
+/// Returns the output directory for the build artifacts.
 fn get_output_dir() -> PathBuf {
     env::var_os("OUT_DIR")
-        .expect("OUT_DIR environment variable not set")
-        .into()
+        .map(PathBuf::from)
+        .expect("OUT_DIR environment variable is not set. Cannot determine output directory.")
 }
 
+/// Checks if the cache is valid by comparing the expected ID with the saved ID.
+#[cfg(not(debug_assertions))]
+fn is_cache_valid(expected_id: &[u8; 32], paths: &[PathBuf; 4]) -> bool {
+    // Check if any required files are missing
+    if paths.iter().any(|path| !path.exists()) {
+        return false;
+    }
+
+    // Attempt to read the saved ID
+    let saved_id = match fs::read(&paths[1]) {
+        Ok(data) => data,
+        Err(_) => return false,
+    };
+
+    expected_id == saved_id.as_slice()
+}
+
+/// Ensures the cache is valid and returns the ELF contents and SP1 Verifying Key.
+#[cfg(not(debug_assertions))]
+fn ensure_cache_validity(program: &str) -> Result<(Vec<u8>, SP1VerifyingKey), String> {
+    let cache_dir = format!("{}/cache", program);
+    let paths = ["elf", "id", "vk", "pk"]
+        .map(|file| Path::new(&cache_dir).join(format!("{}.{}", program, file)));
+
+    // Attempt to read the ELF file
+    let elf = fs::read(&paths[0])
+        .map_err(|e| format!("Failed to read ELF file {}: {}", paths[0].display(), e))?;
+    let elf_hash: [u8; 32] = Sha256::digest(&elf).into();
+
+    if !is_cache_valid(&elf_hash, &paths) {
+        // Cache is invalid, need to generate vk and pk
+        let client = MockProver::new();
+        let (pk, vk) = client.setup(&elf);
+
+        fs::write(&paths[1], elf_hash)
+            .map_err(|e| format!("Failed to write ID file {}: {}", paths[1].display(), e))?;
+
+        fs::write(&paths[2], serialize(&vk).expect("VK serialization failed"))
+            .map_err(|e| format!("Failed to write VK file {}: {}", paths[2].display(), e))?;
+
+        fs::write(&paths[3], serialize(&pk).expect("PK serialization failed"))
+            .map_err(|e| format!("Failed to write PK file {}: {}", paths[3].display(), e))?;
+
+        Ok((elf, vk))
+    } else {
+        // Cache is valid, read the VK
+        let serialized_vk = fs::read(&paths[2])
+            .map_err(|e| format!("Failed to read VK file {}: {}", paths[2].display(), e))?;
+        let vk: SP1VerifyingKey =
+            deserialize(&serialized_vk).map_err(|e| format!("VK deserialization failed: {}", e))?;
+        Ok((elf, vk))
+    }
+}
+
+/// Generates the ELF contents and VK hash for a given program.
 #[cfg(not(debug_assertions))]
 fn generate_elf_contents_and_vk_hash(program: &str) -> (Vec<u8>, [u32; 8], String) {
     // Setup compiler
+    const RISC_V_COMPILER: &str = "/opt/riscv/bin/riscv-none-elf-gcc";
```

Comment:
Is this path also the same in Linux. This feels MacOS-y...

Reply:
Oh yes right! It's not required with the latest version. Created PR #448 to move to latest version.